File: apps/posting/posting_core/verify.py
from __future__ import annotations

import urllib.parse
import urllib.request
from typing import Any

from .clients.bluesky import verify_bluesky_root_visible
from .social_urls import target_public_url


def _http_visible(url: str, timeout: int = 15) -> tuple[bool, str | None]:
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "alexgetman-posting-verify/1.0"})
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            return int(resp.status) < 500, f"http_{resp.status}"
    except Exception as exc:
        return False, str(exc)


def verify_target_record(record: dict[str, Any]) -> dict[str, Any]:
    target = record.get("target")
    status = record.get("status")
    external_id = record.get("external_id")
    url = target_public_url(str(target or ""), external_id=external_id, url=record.get("url"))
    if status != "published":
        return {"target": target, "ok": False, "status": status, "url": url, "reason": record.get("error") or "not_published"}
    if target == "bluesky":
        ok, reason = verify_bluesky_root_visible(external_id)
        return {"target": target, "ok": ok, "status": status, "url": url, "reason": reason}
    if url:
        ok, reason = _http_visible(url)
        return {"target": target, "ok": ok, "status": status, "url": url, "reason": reason}
    return {"target": target, "ok": True, "status": status, "url": url, "reason": "no_public_url_known"}

File: apps/backend/src/ops/verify.ts
import { asc, eq, or } from "drizzle-orm";
import type { BackendDb } from "../db/client.js";
import { posts, postTargets } from "../db/schema.js";

export async function verifyPostTargets(backendDb: BackendDb, ref: string): Promise<Record<string, unknown>[]> {
  const id = Number(ref.replace(/^post:/, ""));
  const post = backendDb.db
    .select({ postKey: posts.postKey })
    .from(posts)
    .where(or(eq(posts.postKey, ref), eq(posts.postId, id), eq(posts.messageId, id)))
    .get();
  if (!post) throw new Error(`post not found: ${ref}`);
  const targets = backendDb.db
    .select({ target: postTargets.target, status: postTargets.status, url: postTargets.url, error: postTargets.error })
    .from(postTargets)
    .where(eq(postTargets.postKey, post.postKey))
    .orderBy(asc(postTargets.target))
    .all();
  return Promise.all(
    targets.map(async (record) => {
      if (record.status !== "published") return { ...record, ok: false, reason: record.error ?? "not_published" };
      if (!record.url) return { ...record, ok: true, reason: "no_public_url_known" };
      try {
        const response = await fetch(record.url, {
          headers: { "user-agent": "alexgetman-backend-verify/1.0" },
          redirect: "follow",
          signal: AbortSignal.timeout(15_000),
        });
        return { ...record, ok: response.status < 500, reason: `http_${response.status}` };
      } catch (error) {
        return { ...record, ok: false, reason: error instanceof Error ? error.message : String(error) };
      }
    }),
  );
}
